File: Codechef/June/Long/ChefAndCities.py
from sys import stdin,stdout
# stdin = open("/Users/seeva92/Workspace/Contests/1.txt","r")
# stdout = open("/Users/seeva92/Workspace/Contests/2.txt","w")
from functools import reduce
from operator import mul
def main():
    # stdout.write("wow")
    N = int(stdin.readline().strip())
    arr = list(map(int,stdin.readline().strip().split()))
    Q = int(stdin.readline().strip())
    for i in range(Q):
        tmp = stdin.readline().strip().split()
        if len(tmp) == 3:
            t,p,q = map(int,tmp)
            arr[p-1] = q
        else:
            t,R = map(int,tmp)
            res = str(reduce(mul,(arr[i] for i in range(0,N,R)),1))
            stdout.write(res[0]+" "+str(int(res)%(1000000007))+'\n')
main()

File: Codechef/June/Long/ChefAndCities.cpp
#include <climits>
#include <cstdio>
#include <iostream>
#include <unordered_map>
#define vi vector<int>
#define all(v) v.begin(), v.end()
#define pii pair<int, int>
#define mp make_pair
#define ll long long
const long long mod = (long long)1e9;
using namespace std;
ll arr[100007];
ll Rarr[100007];
int t, p, f, R, N, Q;
ll query() {
  ll total = 1;
  for (int i = 1; i <= N; i += R) {
    total = (total * arr[i]) % mod;
  }
  return total;
}
int main() {
#ifndef ONLINE_JUDGE
  freopen("/Users/seeva92/Workspace/Contests/1.txt", "r", stdin);
  freopen("/Users/seeva92/Workspace/Contests/2.txt", "w", stdout);
#endif
  ios::sync_with_stdio(false);
  cin.tie(0);
  cin >> N;
  for (int i = 1; i <= N; i++)
    cin >> arr[i];
  cin >> Q;
  for (int i = 0; i < Q; i++) {
    cin >> t;
    if (t == 1) {
      cin >> p >> f;
      arr[p] = f;
    } else {
      cin >> R;
      cout << query() << endl;
    }
  }
}
